tools/chplcheck/src/rules.py

```diff
@@ -535,33 +535,47 @@ def MisleadingIndentation(context: Context, root: AstNode):
         """
         if isinstance(root, Comment):
             return
-
-        prev, prevloop = None, None
+        prev = []
+        fix = None
         for child in root:
             if isinstance(child, Comment):
                 continue
             if might_incorrectly_report_location(child):
                 continue
             yield from MisleadingIndentation(context, child)
 
-            if prev is not None:
-                loc = child.location()
-                prev_loc = prev.location()
-                if loc.start()[1] == prev_loc.start()[1]:
-                    yield AdvancedRuleResult(child, prevloop)
-
-            prev, prevloop = None, None
-            if isinstance(child, Loop) and child.block_style() == "implicit":
-                grandchildren = list(child)
-                # safe to access [-1], loops must have at least 1 child
-                for blockchild in reversed(list(grandchildren[-1])):
-                    if isinstance(blockchild, Comment):
-                        continue
-                    if might_incorrectly_report_location(blockchild):
-                        continue
-                    prev = blockchild
-                    prevloop = child
-                    break
+            if prev and any(
+                p.location().start()[1] == child.location().start()[1]
+                for p in prev
+            ):
+                yield AdvancedRuleResult(child, fix)
+
+            prev = []
+            fix = append_nested_single_stmt(child, prev)
+
+    def append_nested_single_stmt(node, prev: List[AstNode]):
+        if isinstance(node, Loop) and node.block_style() == "implicit":
+            children = list(node)
+            # safe to access [-1], loops must have at least 1 child
+            for blockchild in reversed(list(children[-1])):
+                if isinstance(blockchild, Comment):
+                    continue
+                if might_incorrectly_report_location(blockchild):
+                    continue
+                prev.append(blockchild)
+                append_nested_single_stmt(blockchild, prev)
+                return node  # Return the outermost loop to use an anchor
+        elif isinstance(node, On) and node.block_style() == "implicit":
+            for stmt in node.stmts():
+                if isinstance(stmt, Comment):
+                    continue
+                if might_incorrectly_report_location(stmt):
+                    continue
+                prev.append(stmt)
+                append_nested_single_stmt(stmt, prev)
+                return node  # Return the outermost on to use an anchor
+        # Should we also check for Conditionals here?
+        return None
```

Comment:
It's a little unfortunate that all of this is duplicated. Can you pick out `blockcild` using an `isinstance`, and then leave the `append` in a single place?

```suggestion
        if isinstance(node, Loop) and node.block_style() == "implicit":
            children = list(node)
            # safe to access [-1], loops must have at least 1 child
            inblock = reversed(list(children[-1]))
        elif isinstance(node, On) and node.block_style() == "implicit":
            inblock = node.stmts()
        else:
            # Should we also check for Conditionals here?
            return None
        for stmt in inblock:
            if isinstance(stmt, Comment):
                continue
            if might_incorrectly_report_location(stmt):
                continue
            prev.append(stmt)
            append_nested_single_stmt(stmt, prev)
            return node  # Return the outermost on to use an anchor
        
        # Should we also check for Conditionals here?
        return None
```